### PR TITLE
core+partners/anthropic: Anthropic prompt caching

### PR DESCRIPTION
**Pull request by mrdrprofuroboros**
**Description:** Added support for Anthropic prompt caching, see https://github.com/langchain-ai/langchain/discussions/25625
**Issue:** the issue # it fixes, if applicable
**Dependencies:** bump anthropic>=0.34.0

* just found that it fails the test
will fix it and add usage example to the notebook, so far here:

```python
from langchain_anthropic import ChatAnthropic
from langchain_core.messages import HumanMessage, SystemMessage

model = ChatAnthropic(model="claude-3-5-sonnet-20240620", beta=True)
model.invoke([
    SystemMessage([{
        "type": "text",
        "text": "foo"*1000,
        "cache_control": {"type": "ephemeral"}
    }]),
    HumanMessage("hi!"),
])
```

```
AIMessage(content='Hello! How can I assist you today?', response_metadata={'id': 'msg_01XnYziv7oZtaRw23d45ivSi', 'model': 'claude-3-5-sonnet-20240620', 'stop_reason': 'end_turn', 'stop_sequence': None, 'usage': {'cache_creation_input_tokens': 1500, 'cache_read_input_tokens': 0, 'input_tokens': 9, 'output_tokens': 12}}, id='run-31a51a4f-bb0c-4b24-9c75-d7f7e5f99a89-0', usage_metadata={'input_tokens': 9, 'output_tokens': 12, 'total_tokens': 21, 'cache_creation_input_tokens': 1500, 'cache_read_input_tokens': 0})
```

### CODE REVIEW

**Comment by mrdrprofuroboros** [comment]
@efriis oof, main is running ahead real fast. my changes passed tests / lint before I decided to bump up with "Update branch". Would you mind taking a look and helping to figure out what are the next steps or how can I improve my PR to get it merged?

File: libs/core/langchain_core/messages/ai.py [hunk 1]
```diff
@@ -51,6 +51,10 @@ class UsageMetadata(TypedDict):
     """Count of output (or completion) tokens."""
     total_tokens: int
     """Total token count."""
+    cache_creation_input_tokens: NotRequired[int]
```
Comment (baskaryan) [inline libs/core/langchain_core/messages/ai.py:54]: i dont think we want to add this to core until at least one or two other providers support a similar feature

File: libs/partners/anthropic/langchain_anthropic/chat_models.py [hunk 2]
```diff
@@ -658,6 +666,18 @@ def post_init(cls, values: Dict) -> Dict:
         values["_async_client"] = anthropic.AsyncClient(**client_params)
         return values
 
+    @property
+    def _messages_client(self) -> Messages:
+        if self.beta:
+            return self._client.beta.prompt_caching.messages  # type: ignore[attr-defined]
```
Comment (baskaryan) [inline libs/partners/anthropic/langchain_anthropic/chat_models.py:672]: this feels more specific than just a "beta" flag indicates. are we going to update client to `beta.{x}.messages` every time there's a new beta feature?

also is cache usage not returned if you use the regular client with the beta headers?

**Review by baskaryan — COMMENTED** [review]
thanks for the contribution! i'm very down for including cache token usage in ChatAnthropic outputs but think we'll want to make sure we do it in a future-proof/generalizable way

File: libs/partners/anthropic/langchain_anthropic/chat_models.py (hunk 2 above)
Comment (mrdrprofuroboros) [reply libs/partners/anthropic/langchain_anthropic/chat_models.py:672]: Oh, nice, it actually works. here's an example:
```python
from langchain_anthropic import ChatAnthropic
from langchain_core.messages import HumanMessage, SystemMessage

model = ChatAnthropic(
    model="claude-3-opus-20240229",
    temperature=0,
    extra_headers={"anthropic-beta": "prompt-caching-2024-07-31"}
)

chat = [
    SystemMessage([{
        "type": "text",
        "text": "foo"*1000,
        "cache_control": {"type": "ephemeral"},
    }]),
    HumanMessage("Hi"),
]

model.invoke(chat)
```

returning
```
AIMessage(content='Hello! How can I assist you today?', response_metadata={'id': 'msg_01EuihUPN9JrbzZXuZd6oEu8', 'model': 'claude-3-opus-20240229', 'stop_reason': 'end_turn', 'stop_sequence': None, 'usage': {'input_tokens': 8, 'output_tokens': 12, 'cache_creation_input_tokens': 1500, 'cache_read_input_tokens': 0}}, id='run-a13ecd02-d669-4028-b8a2-56e5113d2417-0', usage_metadata={'input_tokens': 8, 'output_tokens': 12, 'total_tokens': 20})
```

**Comment by huynhphuchuy** [comment]
Could we merge this?

File: libs/partners/anthropic/langchain_anthropic/chat_models.py (hunk 2 above)
Comment (mikevin920) [reply libs/partners/anthropic/langchain_anthropic/chat_models.py:672]: Is there way to include variables in the system prompt but still includes the "cache_control": {"type": "ephemeral"} parameter?

**Pull request closed without merging**